src/bincode/hashmap.zig

```diff
@@ -7,11 +7,113 @@ const hashMapInfo = sig.utils.types.hashMapInfo;
 const FieldConfig = bincode.FieldConfig;
 const Params = bincode.Params;
 
-const readFieldWithConfig = bincode.readFieldWithConfig;
-const readIntAsLength = bincode.readIntAsLength;
-const write = bincode.write;
 const writeFieldWithConfig = bincode.writeFieldWithConfig;
 
+pub fn readCtx(
+    allocator: std.mem.Allocator,
+    comptime H: type,
+    reader: anytype,
+    params: bincode.Params,
+    /// Expects void, or value/namespace with methods (implied `ctx: @TypeOf(ctx)` first parameter):
+    /// * `fn readKey(allocator: std.mem.Allocator, reader: anytype, params: bincode.Params) !Key`, or `pub const readKey = {}`.
+    /// * `fn freeKey(allocator: std.mem.Allocator, key: Key) void`, or `pub const freeKey = {}`.
+    /// * `fn readValue(allocator: std.mem.Allocator, reader: anytype, params: bincode.Params) !Value`, or `pub const readValue = {}`.
+    /// * `fn freeValue(allocator: std.mem.Allocator, value: Value) void`, or `pub const freeValue = {}`.
+    ///
+    /// Void is an indication to utilize default behavior.
+    ctx: anytype,
+) !H {
+    const hm_info = hashMapInfo(H).?;
+
+    const Unmanaged = comptime U: {
+        var unmanaged_info = hm_info;
+        unmanaged_info.management = .unmanaged;
+        break :U unmanaged_info.Type();
+    };
+
+    const ctx_maybe_namespace = if (@TypeOf(ctx) != void) ctx else struct {
+        pub const readKey = {};
+        pub const freeKey = {};
+        pub const readValue = {};
+        pub const freeValue = {};
+    };
+    const ctx_impl = if (@TypeOf(ctx_maybe_namespace) != type) ctx_maybe_namespace else struct {
+        pub fn readKey(
+            _allocator: std.mem.Allocator,
```

Comment:
why are we underscoring params?

Reply:
Because they can't shadow the parameters of the surrounding function; not much else to name them.